static/src/javascripts/projects/common/modules/ui/autoupdate.js

```diff
@@ -0,0 +1,242 @@
+
+                updateTimeoutId = setTimeout(
+                    checkForUpdates,
+                    currentUpdateDelay
+                );
+            } else {
+                // might have been cached so check straight away
+                updateTimeoutId = setTimeout(checkForUpdates, 1);
+            }
+        };
+
+        return fetchJSON(endpoint, {
+            mode: 'cors',
+        })
+            .then(resp => {
+                count = resp.numNewBlocks;
+
+                if (count > 0) {
+                    unreadBlocksNo += count;
+
+                    // updates notification bar with number of unread blocks
+                    mediator.emit('modules:autoupdate:unread', unreadBlocksNo);
+
+                    latestBlockId = resp.mostRecentBlockId;
+
+                    if (isLivePage) {
+                        injectNewBlocks(resp.html);
+
+                        if (scrolledPastTopBlock()) {
+                            toastButtonRefresh();
+                        } else {
+                            displayNewBlocks();
+                        }
+                    } else {
+                        toastButtonRefresh();
+                    }
+                }
+
+                setUpdateDelay();
+            })
+            .catch(() => {
+                setUpdateDelay();
+            });
+    };
+
+    const setUpListeners = () => {
+        bean.on(document.body, 'click', '.toast__button', () => {
+            if (isLivePage) {
+                fastdom.read(() => {
+                    scrollToElement(qwery('.blocks')[0], 300, 'easeOutQuad');
+
+                    fastdom
+                        .write(() => {
+                            $toastButton.addClass('loading');
+                        })
+                        .then(() => {
+                            displayNewBlocks();
+                        });
+                });
+            } else {
+                location.assign(window.location.pathname);
+            }
+        });
+
+        mediator.on('modules:toast__tofix:unfixed', () => {
+            if (isLivePage && unreadBlocksNo > 0) {
+                fastdom
+                    .write(() => {
+                        $toastButton.addClass('loading');
+                    })
+                    .then(() => {
+                        displayNewBlocks();
+                    });
+            }
+        });
+
+        mediator.on('modules:detect:pagevisibility:visible', () => {
+            if (unreadBlocksNo === 0) {
+                revealInjectedElements();
+            }
+
+            currentUpdateDelay = 0; // means please get us fully up to date
+            checkForUpdates();
+        });
+    };
+
+    // init
+    new NotificationCounter().init();
+    new Sticky(toastContainer, {
+        top: options.toastOffsetTop,
+        emitMessage: true,
+        containInParent: false,
+    }).init();
+
+    checkForUpdates();
+    detect.initPageVisibility();
```

Comment:
This pattern is a bit weird. If [`initPageVisibility`](https://github.com/guardian/frontend/blob/master/static/src/javascripts-legacy/lib/detect.js#L309) was called by more than one module on a page, does the event handler get added more than once? We should probably review this at some point